staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go

```diff
@@ -170,8 +192,17 @@ func (o *APIResourceOptions) RunAPIResources() error {
 		errs = append(errs, err)
 	}
 
-	resources := []groupResource{}
+	rl := &metav1.APIResourceList{
```

Comment:
`lists` returned from discovery should satisfy the printobject's requirements?. I mean what about something like this:

```go
	resources := []groupResource{}
	var filteredList []*metav1.APIResourceList

	for _, list := range lists {
		if len(list.APIResources) == 0 {
			continue
		}
		gv, err := schema.ParseGroupVersion(list.GroupVersion)
		if err != nil {
			continue
		}
		apilist := &metav1.APIResourceList{
			TypeMeta: metav1.TypeMeta{
				Kind:       "v1",
				APIVersion: "APIResourceList",
			},
			GroupVersion: gv.String(),
		}
		var apiresources []metav1.APIResource
		for _, resource := range list.APIResources {
			if len(resource.Verbs) == 0 {
				continue
			}
			// filter apiGroup
			if o.groupChanged && o.APIGroup != gv.Group {
				continue
			}
			// filter namespaced
			if o.nsChanged && o.Namespaced != resource.Namespaced {
				continue
			}
			// filter to resources that support the specified verbs
			if len(o.Verbs) > 0 && !sets.NewString(resource.Verbs...).HasAll(o.Verbs...) {
				continue
			}
			// filter to resources that belong to the specified categories
			if len(o.Categories) > 0 && !sets.NewString(resource.Categories...).HasAll(o.Categories...) {
				continue
			}

			resources = append(resources, groupResource{
				APIGroup:        gv.Group,
				APIGroupVersion: gv.String(),
				APIResource:     resource,
			})
			apiresources = append(apiresources, resource)
		}
		if len(apiresources) > 0 {
			apilist.APIResources = apiresources
			filteredList = append(filteredList, apilist)
		}
	}

	// if output format is json or yaml, directly print filteredList and short cut

	// if output format is default, print resources as always

	if o.NoHeaders == false && o.Output != "name" {
```

Reply:
> `lists` returned from discovery should satisfy the printobject's requirements?

I don't think so. And that's why I went with the complicated looking approach. `lists` is a slice of type `APIResourceList`. While `APIResourceList` itself implements the `runtime.Object` interface, its slice `[]*metav1.APIResourceList` doesn't.

Reply:
@ardaguclu I have implemented it the way you suggested in this comment. Look at 4e52bc3. IMO, the problem with this approach is that it prints a bunch of JSON object not a unified JSON object.

So in the next commit 562d48a I flattened the `filteredList` and print a unified JSON object. However, it doesn't have the `groupVersion` info that we presently print under the `APIVersion` column.

This is the reason I spent too long before opening the PR. In hindsight, I feel that I should have brought it up earlier.

Note that there are no tests yet. I would like to get feedback on what we want the output to be like, and then I'll focus on adding tests.

cc @brianpursley